### PR TITLE
Add link parameter to offer template

### PR DESCRIPTION
**Pull request by jparyani**
This allows you to make the whole offer template into a link. At the
moment, the only thing you can specify is a scheme, and the rest is
autogenerated for you.

This needs some input on what we're actually going to do about filtering dangerous schemes. Either add a whitelist or only allow them to start with "sandstorm-" (or automatically prefix it ourselves).

### CODE REVIEW

File: shell/client/grain-client.js [hunk 1]
```diff
@@ -1509,12 +1523,18 @@ Meteor.startup(function () {
         const renderedTemplate = template.replace(/\$API_TOKEN/g, tokenId)
                                          .replace(/\$API_HOST/g, host)
                                          .replace(/\$GRAIN_TITLE_SLUG/g, grainTitleSlug);
+        let link = undefined;
+        if (rawLink) {
+          link = `${rawLink.scheme}://${window.location.protocol}//${host}#${tokenId}`;
```
Comment (kentonv) [inline shell/client/grain-client.js]: Change the first `://` to just `:` -- the `//` doesn't make sense since it's not followed by a hostname.

**Comment by kentonv** [comment]
How about `clientapp-*:` for the scheme?

I wonder if we should use basic-auth-like syntax instead of our webkey syntax...

    clientapp-foo:https://*:<token>@<host>/

Apps that know they support `Authorization: bearer` would of course be free to use the password as a token. (Note that the URL spec allows `*` as a username but this also seems like a good way to indicate "doesn't matter".)

**Comment by jparyani** [comment]
Discussed in person with @kentonv, but using basic auth doesn't work on android. I've tracked it down to the "@" character causing problems. It results in an ERR_UNKNOWN_URL_SCHEME for the android browser.

**Comment by kentonv** [comment]
The test failures appear to be legitimate, BTW.

**Comment by jparyani** [comment]
Ready to merge

File: shell/client/grain-client.js [hunk 2]
```diff
@@ -1428,7 +1430,17 @@ Meteor.startup(function () {
           unauthenticated: Match.Optional(Object),
           // Note: `unauthenticated` will be validated on the server. We just pass it through
           //   here.
+          link: Match.Optional({
```
Comment (kentonv) [inline shell/client/grain-client.js]: Let's call this "clientapp".

File: shell/public/offer-template.html [hunk 3]
```diff
@@ -129,7 +140,14 @@
        } else {
           var errorText = "Error refreshing token. Reloading the page might help.";
           console.error(xhr.responseText);
-          document.getElementById("text").textContent = errorText;
+          var textElement = document.getElementById("text");
+          textElement.textContent = errorText;
+          var link = document.getElementById("href");
```
Comment (kentonv) [inline shell/public/offer-template.html]: It's the id "link"?

File: shell/client/grain-client.js [hunk 4]
```diff
@@ -1428,7 +1430,17 @@ Meteor.startup(function () {
...
+            scheme: String,
+          }),
         });
+
+        rawLink = call.link;
+        // rfc3986 specifies schemes as:
+        // scheme = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+        if (rawLink && (rawLink.scheme.search(/[^a-zA-Z0-9+-.]/) !== -1)) {
```
Comment (kentonv) [inline shell/client/grain-client.js]: Use Match.Where

**Pull request merged**